### PR TITLE
[MLv2] Adjust filter clause display-names

### LINKED ISSUE #34332: [MLv2] BE - Fix filter display names
Epic #34112

Filter display names coming from MLv2 are currently different from what we have with MLv1.

Please use a frontend test suite on the `mlv2-filter-name-tests` branch to see the differences
You can run that test suite using the following command:

```
yarn jest frontend/src/metabase-lib/filterNames.unit.spec.ts
```

FYI, the test suite currently doesn't contain tests for:

- time filters
- lat/lng `"inside"` filters

### PR DESCRIPTION
**Pull request by snoe**
Fixes #34332

### CODE REVIEW

File: src/metabase/lib/filter.cljc [hunk 1]
```diff
@@ -59,45 +63,162 @@
...
+      (i18n/tru "{0} is equal to {1}" (->display-name a) (->display-name b))
+
+      [:= _ (a :guard temporal?) (b :guard (some-fn int? string?))]
+      (i18n/tru "{0} is on {1}" (->display-name a) (->temporal-name a b))
+
+      [:!= _ (a :guard numeric?) b]
+      (i18n/tru "{0} is not equal to {1}" (->display-name a) (->display-name b))
+
+      [:!= _ (a :guard (unit-is :day-of-week)) (b :guard (some-fn int? string?))]
+      (i18n/tru "{0} excludes {1}" (->unbucketed-display-name a) (inflections/plural (->temporal-name a b)))
```
Comment (metamben) [inline src/metabase/lib/filter.cljc:100]: I wonder how this can be translated to other languages.

File: src/metabase/lib/filter.cljc [hunk 2]
```diff
@@ -59,45 +63,162 @@
...
+      (i18n/tru "{0} is greater than or equal to {1}" (->display-name x) (->display-name y))
+
+      [:starts-with _ x (y :guard string?)]
+      (i18n/tru "{0} starts with {1}"                 (->display-name x) y)
+
+      [:starts-with _ x y]
+      (i18n/tru "{0} starts with {1}"                 (->display-name x) (->display-name y))
+
+      [:ends-with _ x (y :guard string?)]
+      (i18n/tru "{0} ends with {1}"                   (->display-name x) y)
```
Comment (metamben) [inline src/metabase/lib/filter.cljc:172]: Why not define `->display-name` for strings as identity?

File: src/metabase/lib/util.cljc [hunk 3]
```diff
@@ -55,6 +57,23 @@
...
+
+(defn original-isa?
+  "Returns whether the type of `expression` isa? `typ`.
+   If the expression has an original-effective-type due to bucketing, check that."
+  [expression typ]
+  (isa?
+    (if-let [expr-typ (and (clause? expression)
+                           (:metabase.lib.field/original-effective-type (second expression)))]
+      expr-typ
+      (lib.schema.expression/type-of expression))
```
Comment (metamben) [inline src/metabase/lib/util.cljc:74]:
```suggestion
    (or (and (clause? expression)
             (:metabase.lib.field/original-effective-type (second expression)))
        (lib.schema.expression/type-of expression))
```

File: test/metabase/lib/filter_test.cljc [hunk 4]
```diff
@@ -433,3 +433,240 @@
...
+       {:clause [:!= fk 1 2 12], :name "User ID is not 3 selections"}
+       {:clause [:> fk 1], :name "User ID is greater than 1"}
+       {:clause [:< fk 1], :name "User ID is less than 1"}
+       {:clause [:between fk 1 10], :name "User ID is between 1 and 10"}
+       {:clause [:>= fk 1], :name "User ID is greater than or equal to 1"}
+       {:clause [:<= fk 1], :name "User ID is less than or equal to 1"}
+       {:clause [:is-null fk], :name "User ID is empty"}
+       {:clause [:not-null fk], :name "User ID is not empty"}])))
+
+(deftest ^:parallel -frontend-filter-display-names-test
```
Comment (metamben) [inline test/metabase/lib/filter_test.cljc:546]:
```suggestion
(deftest ^:parallel string-frontend-filter-display-names-test
```

**Review by metamben — APPROVED** [review]
LGTM

File: frontend/src/metabase/query_builder/components/view/QuestionFilters.jsx [hunk 5]
```diff
@@ -51,30 +54,37 @@ export default function QuestionFilters({
...
-              sizeToFit
-            >
-              <FilterPopover
-                isTopLevel
-                query={query}
-                filter={filter}
-                onChangeFilter={newFilter =>
-                  onQueryChange(newFilter.replace().rootQuery())
+          filters.map((filter, index) => {
+            const v2Filter = Lib.filters(v2Query, -1)[index];
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/QuestionFilters.jsx:58]: This is not correct! `topLevelFilters` returns filters for the last 2 stages in the query

**Review by ranquild — CHANGES_REQUESTED** [review]
I strongly believe that in this PR there should be no FE changes. We will update affected E2E tests in the FE integration branch. Some FE changes in this PR are not correct and cannot be easily fixed.

Please remove FE changes and I'll approve the PR.

**Review by ranquild — APPROVED** [review]

File: shared/src/metabase/shared/util/internal/time.clj [hunk 6]
```diff
@@ -203,3 +203,102 @@
   "Returns the time elapsed between `before` and `after` in days (an integer)."
   [before after]
   (.toDays (t/duration before after)))
+
+(defn- matches-time? [input]
+  (re-matches #"\d\d:\d\d(?::\d\d(?:\.\d\d\d?)?)?" input))
```
[the remark this answers is not in the document]
Comment (snoe) [reply shared/src/metabase/shared/util/internal/time.clj:208]: I changed the regex below to be `\d+` missed this one, thank you!

File: src/metabase/lib/filter.cljc (hunk 2 above)
Comment (snoe) [reply src/metabase/lib/filter.cljc:172]: String as identity would affect custom expressions that would have a larger chance to be ambiguous between columns and strings. I didn't want to expand the scope more than we needed.

**Pull request merged**